### ISSUE TITLE
ADIOS 2.6.0: Linker issue with DataManSerializer

### ISSUE THREAD

**Issue by ax3l** [ornladios/ADIOS2#2494]
**Describe the bug**
Since at least two weeks ([history here](https://dev.azure.com/axelhuebl/openPMD-api/_build?definitionId=1&_a=summary)), I have problems building ADIOS2 from Spack, when deselecting both MPI and Fortran support.

**To Reproduce**
See full nightly integration tests here: https://dev.azure.com/axelhuebl/openPMD-api/_build?definitionId=1&_a=summary
Recipe: https://github.com/openPMD/openPMD-api/blob/a88603221062de7dd240b78a51bfe9332e139857/azure-pipelines.yml#L247-L276

Arch: x86_64 (e.g. broadwell)
Ubuntu: 18.04
GCC: 10.1.0
CMake: 3.18.4

Spack command: `spack install -v --fail-fast openpmd-api -mpi ^adios2~fortran` (latest Spack)

```
[ 90%] Linking CXX executable ../../../../../bin/sst_conn_tool
cd /tmp/vsts/spack-stage/spack-stage-adios2-2.6.0-irroopva45fls4anpem6prf53ivjwfea/spack-build-irroopv/source/adios2/toolkit/sst/util && /home/vsts/work/1/s/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-10.1.0/cmake-3.18.4-xrlaq3ksiobjwlnqdkx2tqhjpm54dc2a/bin/cmake -E cmake_link_script CMakeFiles/sst_conn_tool.dir/link.txt --verbose=1
/home/vsts/work/1/s/spack/lib/spack/env/gcc/g++ -O3 -DNDEBUG CMakeFiles/sst_conn_tool.dir/sst_conn_tool.c.o CMakeFiles/sst_conn_tool.dir/sst_conn_tool.cxx.o ../CMakeFiles/sst.dir/dp/dp.c.o ../CMakeFiles/sst.dir/dp/evpath_dp.c.o ../CMakeFiles/sst.dir/cp/cp_reader.c.o ../CMakeFiles/sst.dir/cp/cp_writer.c.o ../CMakeFiles/sst.dir/cp/cp_common.c.o ../CMakeFiles/sst.dir/cp/ffs_marshal.c.o ../CMakeFiles/sst.dir/sst_comm.cpp.o ../CMakeFiles/sst.dir/dp/rdma_dp.c.o ../CMakeFiles/sst.dir/cp/ffs_zfp.c.o -o ../../../../../bin/sst_conn_tool  -Wl,-rpath,/tmp/vsts/spack-stage/spack-stage-adios2-2.6.0-irroopva45fls4anpem6prf53ivjwfea/spack-build-irroopv/lib:/home/vsts/work/1/s/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-10.1.0/libfabric-1.11.0-e2vjq7qwi3zopdzj7iv6bsksv4abj5nv/lib:/home/vsts/work/1/s/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-10.1.0/zfp-0.5.5-2vnclarae53txqolun2q3hdll3m7hj47/lib: ../../../../../lib/libadios2_core.so.2.6.0 -lpthread ../../../../../lib/libadios2_evpath.so ../../../../../lib/libadios2_ffs.so.1.6.0 ../../../../../lib/libadios2_atl.so.2.2.1 ../../../../../lib/libadios2_taustubs.so /home/vsts/work/1/s/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-10.1.0/libfabric-1.11.0-e2vjq7qwi3zopdzj7iv6bsksv4abj5nv/lib/libfabric.so /home/vsts/work/1/s/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-10.1.0/zfp-0.5.5-2vnclarae53txqolun2q3hdll3m7hj47/lib/libzfp.so.0.5.5 -Wl,-rpath-link,/tmp/vsts/spack-stage/spack-stage-adios2-2.6.0-irroopva45fls4anpem6prf53ivjwfea/spack-build-irroopv/lib 
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<long>(long*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<unsigned short>(unsigned short*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<float>(float*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<signed char>(signed char*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<long double>(long double*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<short>(short*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<double>(double*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<int>(int*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<unsigned char>(unsigned char*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<unsigned long>(unsigned long*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<std::complex<float> >(std::complex<float>*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<unsigned int>(unsigned int*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
../../../../../lib/libadios2_core.so.2.6.0: undefined reference to `int adios2::format::DataManSerializer::GetData<std::complex<double> >(std::complex<double>*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&)'
collect2: error: ld returned 1 exit status
```

**Expected behavior**
A successful install.

**Desktop (please complete the following information):**
 - OS/Platform: Ubuntu 18.04 on Azure Pipelines
 - Build: GCC 10.1.1, no-fortran, no-MPI build

**Additional context**
N/A

**Following up**
N/A

**Comment by ax3l** [comment]
@chuckatkins do you have an idea on this?

**Comment by ax3l** [comment]
@chuckatkins @pnorbert I can reproduce this locally on ADIOS 2.6.0 from Spak with GCC 10.2.0 on Ubuntu 20.04 even with MPI.

**Comment by ax3l** [comment]
@pnorbert's work-around: DataMan is pulled optionally (with ZeroMQ), so disabling that can be a work-around: `spack install adios2 ~dataman`

**Comment by williamfgc** [comment]
Duplicate #2457

**Comment by ax3l** [comment]
I think I will disable dataman's default "on" in Spack for now - this impacts us too much in production.

@chuckatkins a nightly integration test that tests all installation methods would be good to catch this :) Something like this: https://github.com/openPMD/openPMD-api/tree/azure_install

**Comment by pnorbert** [comment]
I think this is fixed with master and is only a problem with 2.6.0.  On my Ubuntu 18.04 and with gcc 10.2.0, I got spack reproduce this error for 2.6.0 but built master fine.
`$ spack install adios2 ~mpi ~bzip2 ~blosc ~sst ~png ~sz ~zfp% gcc@10.2.0` -- failed
`$ spack install adios2@master ~mpi ~bzip2 ~blosc ~sst ~png ~sz ~zfp% gcc@10.2.0` -- built

**Comment by ax3l** [comment]
Thanks for testing this - do you recall which PR contains the fix?
It might be worth releasing a 2.6.1 release or patching the 2.6.0 release in Spack and other package mangers after the fact.

**Comment by ArchangeGabriel** [comment]
I can confirm this is fixed in 2.7.x, so maybe this should be closed now that a fixed release is available. ;)

**Comment by ax3l** [comment]
Sound good, thanks!
